src/v/kafka/client/test/data_queue_test.cc

```diff
@@ -198,6 +211,58 @@ TEST(DataQueueTest, TestEdgeCases) {
     ASSERT_TRUE(queue.can_insert(1024));
 }
 
+// the queue as written will indefinitely starve out large operations in
+// exchange for smaller ones, to fix we probably need a fair condition_variable
+TEST_CORO(DataQueueTest, QueueUnfairness) {
+    // constants for making fetched_topic_data
+    static constexpr size_t small_size = 128u;
+    static constexpr auto make_small = +[] { return make_data(1, small_size); };
+    static constexpr size_t big_size = 2048;
+
+    data_queue queue(max_bytes, max_count);
+
+    // drop a small record in the queue
+    co_await queue.push(make_small(), small_size);
+
+    // attempt to push an oversized record, and keep tabs on its successful push
+    bool big_push_has_completed{false};
+    auto big_push = queue.push(make_data(1, big_size), big_size)
+                      .then([&big_push_has_completed] {
+                          big_push_has_completed = true;
+                          return ss::now();
+                      });
+
+    // float around the task queue a bit
+    co_await yield_a_few_times(10);
```

Comment:
Maybe you want https://github.com/redpanda-data/redpanda/blob/00e787123dec1887634efe678b4c8fb62199adce/src/v/test_utils/async.h#L131? You still must be careful when using it.

in debug mode in seastar the task queue is shuffled, so this is possible to yield flakes: https://github.com/redpanda-data/seastar/blob/548c9c162870be0a6049b5f030ed56bdb6f65f6e/src/core/reactor.cc#L2784